isa/kernel.org/instruction-set.rst

```diff
@@ -16,77 +30,141 @@ The eBPF calling convention is defined as:
  * R6 - R9: callee saved registers that function calls will preserve
  * R10: read-only frame pointer to access stack
 
-R0 - R5 are scratch registers and eBPF programs needs to spill/fill them if
-necessary across calls.
+Registers R0 - R5 are scratch registers, meaning the BPF program needs to either
+spill them to the BPF stack or move them to callee saved registers if these
+arguments are to be reused across multiple function calls. Spilling means
+that the value in the register is moved to the BPF stack. The reverse operation
+of moving the variable from the BPF stack to the register is called filling.
+The reason for spilling/filling is due to the limited number of registers.
+
+*Linux implementation note*: In the Linux kernel, the exit value for eBPF
+programs is passed as a 32 bit value.
+
+Upon entering execution of an eBPF program, register R1 initially contains
```

Comment:
eBPF for Windows does the same.

So does generic-ebpf (https://github.com/generic-ebpf/generic-ebpf/blob/dev/sys/dev/ebpf/ebpf_interpreter.c#L35)

@Alan-Jowett it looks like uBPF uses both R1 and R2 (https://github.com/iovisor/ubpf/blob/master/vm/ubpf_vm.c#L207) is that right?

Reply:
IIRC, you can use up to 5 arguments with uBPF if you JIT (the link above is for the uBPF interpreter).  I need to go back and take a look at that again though, it's been a while since I did the NVMe eBPF prototyping.

Reply:
uBPF is inconsistent here. For the interpreter, it passes 2 parameters, but for the JIT it only handles one.

Reply:
rbpf uses R1 only, for that matter.

Reply:
Ok, updated text, please check.

Reply:
@dthaler - the new text looks good to me.

Just to follow up w/ @Alan-Jowett because I want to make sure I understand the comment about uBPF JIT only handling 1 argument.

The ubpf_jit_fn typedef takes two parameters - void *mem and size_t mem_len.  But it can support more than that.

Take the following program:
```
#include <stdint.h>

uint32_t
test_prog(uint64_t arg1, uint64_t arg2, uint64_t arg3, uint64_t arg4, uint64_t arg5)
{
        /* GNU/Linux x86 calling convention uses RCX for the 4th argument, but
         * ubpf maps x86 R9 register to eBPF R4 register, so arg4 doesn't work
         * correctly.  Add the remaining arguments instead and see if the ubpf JIT
         * handles them correctly.
         */
        return arg1 + arg2 + arg3 + arg5;
}
```

and compile it:
```
% clang -target bpf -O2 -c bpf.c bpf.o
% llvm-objdump -d bpf.o

bpf.o:  file format ELF64-BPF

Disassembly of section .text:
0000000000000000 test_prog:
       0:       bf 20 00 00 00 00 00 00         r0 = r2
       1:       0f 10 00 00 00 00 00 00         r0 += r1
       2:       0f 30 00 00 00 00 00 00         r0 += r3
       3:       0f 50 00 00 00 00 00 00         r0 += r5
       4:       95 00 00 00 00 00 00 00         exit
```

Now use this ubpf test wrapper.  Note that I've created a new function signature new_jit_fn that takes 5 parameters instead of 2.

```
#include "ubpf.h"
#include <stdlib.h>

typedef uint32_t (*new_jit_fn)(uint64_t arg1, uint64_t arg2, uint64_t arg3, uint64_t arg4, uint64_t arg5);

int main(int argc, char **argv)
{
        FILE *file = fopen(argv[1], "r");
        char data[16 * 1024];
        size_t rc, offset = 0;
        struct ubpf_vm *vm = ubpf_create();
        char *errmsg;
        new_jit_fn fn;
        uint32_t ret;

        while ((rc = fread(data + offset, 1, sizeof(data) - offset, file)) > 0) {
                offset += rc;
        }

        ubpf_load_elf(vm, data, offset, &errmsg);
        fn = (new_jit_fn)ubpf_compile(vm, &errmsg);
        ret = fn(10, 20, 30, 40, 50);
        printf("ret = %d\n", ret);

        return 0;
}
```

Now build it, and run it against the bpf program:

```
% cc run.c -I/home/jim/git/ubpf/vm/inc -L/home/jim/git/ubpf/vm -lubpf -o run --static
% ./run bpf.o
ret = 110
```

I haven't worked out exactly why ubpf maps R9 instead of RCX to eBPF R4.  If I apply this patch, the 4th argument works as well:

```
diff --git a/vm/ubpf_jit_x86_64.c b/vm/ubpf_jit_x86_64.c
index 1006c75..03eb2a2 100644
--- a/vm/ubpf_jit_x86_64.c
+++ b/vm/ubpf_jit_x86_64.c
@@ -141,6 +141,8 @@ translate(struct ubpf_vm *vm, struct jit_state *state, char **errmsg)
         emit_mov(state, platform_parameter_registers[0], map_register(1));
     }

+    emit_mov(state, RCX, map_register(4));
+
     /* Copy stack pointer to R10 */
     emit_mov(state, RSP, map_register(10));

```